### ISSUE TITLE
DX11 Media Foundation / DX12 Video Decoding Plans

### ISSUE THREAD

**Issue by h1cks** [amerkoleci/Vortice.Windows#103]
Are there plans to look at looking at these 2 sets of APIs for Vortice.Windows.

If not, happy to look at alternatives.   Great work on the implementation btw.

**Comment by amerkoleci** [comment]
Of course yes, as described in readme new proposal goes through sponsor part, i will gladly implement it if you plan to somehow sponsor my profile, it's spare time project and need to allocate time for it.

**Comment by h1cks** [comment]
Yes, was thinking of sponsoring you.  Appreciate the effort.

**Comment by amerkoleci** [comment]
Don't feel forced to do so :) was just telling you the situation, thanks for bringing some feature requests as well

**Comment by h1cks** [comment]
I don't feel forced, hehhe  :D  Appreciate it, this API is the only DX 11 C# implementation going forward.  I work hard on my own game and personally feel my own pressure that as the .NET platform goes forward.  This API gives me the chance to work on my projects going forward (without having to jump to c++!)

**Comment by amerkoleci** [comment]
First of all, thanks for sponsoring me, second I'm working on C# game engine/framework with focus on net5 core and future net6 (https://github.com/amerkoleci/alimer_sharp)

**Comment by h1cks** [comment]
Not a problem on that front.  You've done a ton of work on the platform since Sharpdx was closed off.  

My game project can be found here [Youtube channel](https://www.youtube.com/channel/UCQgOLROabm_1W6NIsa0koBw) and https://insanesoftware.itch.io/pacific-strafe-alpha

I will fork your project and have a look at a couple of things.  DirectInput although old, is something I still need.  I might have a play and see if i can set up bindings for my own learning.

**Comment by amerkoleci** [comment]
Wow, that game looks sick, you've done all by yourself? editor etc?

**Comment by h1cks** [comment]
Yeah, started in june 2016, from scratch.  The whole game engine is my own, as well as the tooling, I wrote a level editor and object editor as well. Use Blender for the model creation.  Some of the model donated to me by some very generous modellers.

I moved to the SDK type builds in VS 2019 and was able to update to .NET 5.0 still while using Sharpdx 4.2.0.  Performance improvement on the CPU side was pretty impressive, hence my interest in bringing it to .Net 5/6 with c# 9.0.

As least you can see how much I will use vortice ><

**Comment by amerkoleci** [comment]
Yeah, I plan to move vortice.windows to target net5 soon and remove netstandard 2.0 support.

**Comment by amerkoleci** [comment]
Direct3D12 Video bindings just landed and Direct3D12 has been moved to target net5 now.

**Comment by h1cks** [comment]
@amerkoleci has there been any consideration on moving the older DX11 Media Foundation.  I know this is a substantial piece of work, I'm sort of left with no video alternatives atm.  Keep up the good work.

**Comment by h1cks** [comment]
Just reopened issue for media foundation build failing, I know busy at the moment @amerkoleci.  appreciate the work.

I've spent some time updating my project from my Sharpdx to Vortice.  Only having to update deltas.  the amount of work you have put in has made it so much easier to migrate.

**Comment by h1cks** [comment]
@amerkoleci is the current media foundation build breaking?   I looked at the build and seems to be.  Is this to be another Nuget package that is to be published.  Good work btw...also starting tweeting Vortice windows out to get more awareness of this project.

**Comment by amerkoleci** [comment]
Yes, currently is in preview state, https://github.com/amerkoleci/Vortice.Windows/blob/main/src/Vortice.MediaFoundation/Vortice.MediaFoundation.csproj#L10

If it works I can enable nuget package generation?

**Comment by h1cks** [comment]
@amerkoleci assume this will expose the fmv playback components?

Also, finding some random crashing in release that isnt occurring in debug with my game. Due to the size of my project might take sometime to find issues.  Will continue to test vortice against my game.  And diagnose if the issue is within my game first.

**Comment by h1cks** [comment]
> Yes, currently is in preview state, https://github.com/amerkoleci/Vortice.Windows/blob/main/src/Vortice.MediaFoundation/Vortice.MediaFoundation.csproj#L10
> 
> If it works I can enable nuget package generation?

I will create a test project in vortice and start testing this on the weekend.

**Comment by SuRGeoNix** [comment]
👍 Good job @amerkoleci just replaced also SharpDX with Vortice on my [media player library](https://github.com/SuRGeoNix/Flyleaf). Took less than an hour!

**Comment by amerkoleci** [comment]
@SuRGeoNix Nice one, spread the word then :)

**Comment by amerkoleci** [comment]
Preview nuget package for Media Foundation just landed on nuget: https://www.nuget.org/packages/Vortice.MediaFoundation/

**Comment by h1cks** [comment]
> 👍 Good job @amerkoleci just replaced also SharpDX with Vortice on my [media player library](https://github.com/SuRGeoNix/Flyleaf). Took less than an hour!

Would love to know more about your migration, what components did you use.  On my current Sharpdx project im using the older media engine components.

**Comment by h1cks** [comment]
@amerkoleci you're getting pretty close to feature complete for now, assume support for future DX12 components etc will come?

**Comment by amerkoleci** [comment]
Yes, I plan to add D3D12 support as well, consider that d3d12video.h is supported under Vortice.Direct3D12.

**Comment by h1cks** [comment]
I've been tweeting your updates so there can be more coverage.  Seeing that Sharpdx hasn't had an update in 2+ years, it's getting critical for a new path be in place for devs on a supported library.

**Comment by amerkoleci** [comment]
Which is your Twitter account?

**Comment by h1cks** [comment]
https://twitter.com/software_insane

**Comment by h1cks** [comment]
I just put out notes on my release of my game, its fairly new account.

Seeing I have my steam support now working, I will be doing more around getting my game on steam and into a releasable condition.  The twitter account will be useful going forward.

**Comment by h1cks** [comment]
@amerkoleci will close this ticket when I have created a sample to add to Vortice.  At the moment flat out on various projects, but keen to commit a sample for this.

**Comment by h1cks** [comment]
@amerkoleci  noticed you have added code around Source Resolver for media creation.  Current release does not include the new code you have written.

I've tried to pull the MediaSource object from the Media pointer after creating an object as per below using the documentation from MS -> https://docs.microsoft.com/en-us/windows/win32/medfound/using-the-source-resolver

Can safely say my approach is wrong, because trying to query the interface for the MediaSource object fails with an interface does not exist error.  Can you direct me here a little, Im just following the Microsoft tutorial.

```
if (MediaFactory.MFCreateSourceResolver(out IMFSourceResolver resolver_).Success)
{
    resolver_.CreateObjectFromURL(Url, 2, null, out ObjectType type_, out IUnknown object_);

    IMFMediaSource source_ = (object_ as ComObject).QueryInterface<IMFMediaSource>();
}

```

**Comment by amerkoleci** [comment]
You can do this:
IMFMediaSource source_ = new ComObject(object_).QueryInterface<IMFMediaSource>();

**Comment by amerkoleci** [comment]
You can use ComObject.As<T>(), as well  QueryInterface etc

**Comment by h1cks** [comment]
Cheers, but i'm still getting a no such interface supported error.  I must be doing the execution of this wrong, it is though returnnig a com object, just doesn't contain the related interface.

**Comment by amerkoleci** [comment]
Sad to hear that, with new release https://github.com/amerkoleci/Vortice.Windows/commit/efb03fecbd99453695dfe923c3fd7fc1e78232bd you will be able to do:

```
string url = "ergon.wav";
            using (IMFSourceResolver resolver = MFCreateSourceResolver())
            {
                var test = File.Exists(url);
                using var byteStream = resolver.CreateObjectFromURLAsByteStream(url);
                using var mediaSource = resolver.CreateObjectFromURL(url);
                Console.WriteLine("");
            }

```

**Comment by h1cks** [comment]
> Sad to hear that, with new release [efb03fe](https://github.com/amerkoleci/Vortice.Windows/commit/efb03fecbd99453695dfe923c3fd7fc1e78232bd) you will be able to do:
> 
> ```
> string url = "ergon.wav";
>             using (IMFSourceResolver resolver = MFCreateSourceResolver())
>             {
>                 using var byteStream = resolver.CreateObjectFromURLAsByteStream(url);
>                 using var mediaSource = resolver.CreateObjectFromURL(url);
>                 Console.WriteLine("");
>             }
> ```

Will wait till the new version is out!

I'm implementing the renderer from this step by step from Microsoft -> https://docs.microsoft.com/en-us/windows/win32/medfound/how-to-play-unprotected-media-files

**Comment by amerkoleci** [comment]
I just enabled nuget releases (beta version) during push, @h1cks you can grab the latest 2.0.19-beta version

**Comment by h1cks** [comment]
Took a little while to show in nuget but see them  now and will pull.  will do a quick once over to see if there are any issues in general, i will be working again on FMV this weekend when I get time.  At the moment I'm finishing some more changes to my game (time of day lighting and water rendering changes).  Once done, will get back to video playback.

**Comment by h1cks** [comment]
Update: A few function name changes and some object no longer needing to be disposed.  But all good on running my game.  I've also had a quick look at the video playback and the IMFResolver object is now being returned, I can now proceed with the rest of the FMV dev work. Thanks!

**Comment by andrew-boyarshin** [comment]
@h1cks
> some object no longer needing to be disposed

You just removed some `using`s? That's likely a memory leak (until the next SharpGen version).

If you had sth like this:
<https://github.com/amerkoleci/Vortice.Windows/blob/bb4a43429b0e1ea58d3762a794845653e9fee0e0/src/Vortice.Dxc/DxcCompiler.cs#L19-L22>

You need now to do sth like this:
<https://github.com/amerkoleci/Vortice.Windows/blob/7591f72bac51728a7bcb19848572ba33120053b1/src/Vortice.Dxc/DxcCompiler.cs#L20-L28>

**Comment by amerkoleci** [comment]
> @h1cks
> 
> > some object no longer needing to be disposed
> 
> You just removed some `using`s? That's likely a memory leak (until the next SharpGen version).
> 
> If you had sth like this:
> https://github.com/amerkoleci/Vortice.Windows/blob/bb4a43429b0e1ea58d3762a794845653e9fee0e0/src/Vortice.Dxc/DxcCompiler.cs#L19-L22
> 
> You need now to do sth like this:
> https://github.com/amerkoleci/Vortice.Windows/blob/7591f72bac51728a7bcb19848572ba33120053b1/src/Vortice.Dxc/DxcCompiler.cs#L20-L28

This has to be fixed probably, using and IDisposable pattern was cool, this will lead to lot of memory leaks.

**Comment by h1cks** [comment]
@amerkoleci funnily enough it was the vertex buffer bindings that no longer are IDisposable.  And that the SetVertexBuffers was changed to SetVertexBuffer

All good.

**Comment by amerkoleci** [comment]
I'm pushing packages to github now, can you try to use them from: https://github.com/amerkoleci?tab=packages
Stable packages goes to nuget as before.

**Comment by h1cks** [comment]
Push anyway, you have a number of other fixes that should go out. and if the video playback wasnt working/complete before, then it wont affect the the new release.

**Comment by Gawson** [comment]
Hi @amerkoleci ,
First of all great job with bringing working DX wrapper after SharpDX death.
I've been watching this thread and discussion about Media Foundation support in Vortice. I know that this is beta but I'm curious if I'm doing something wrong or IAsyncCallback is not usable in current state? I have no clue how to implement such callback even after digging into MS docs and my own old samples with SharpDX.
Maybe this is not the best thread to start discussion but if this is in preliminary state there is no sense to create an issue. If it should work maybe my skills are not enough or I do not understand the idea behind interfaces generated in the way the are generated.

**Comment by h1cks** [comment]
@Gawson the wrapper for vortice is possibly not working.  Due to the size of the interface, there is most likely some issues.  Especially in mediafoundation which I have been working on.  I am still testing AsyncCallback at the moment.

I have been working on a sample from the MS code, the link to my mediafoundation sample is here
https://github.com/h1cks/Vortice.Windows

You should see a hellomediafoundation example. I am also working on mappings, I should have some time today.   I have most of the set up working until I execute BeginEvent, but im still not completely familiar with the AsyncCallback interface, so I havent had time to study what is missing.

@amerkoleci I see that if I do the mediaengine interface, I basically have to #include ALL projects in Vortice.  Will continue to look at getting media foundation working.

**Comment by amerkoleci** [comment]
@h1cks good job, keep me updated, if you have any PR just send and I'll review.
@Gawson Same, if you have trouble ping me with specific question.

**Comment by h1cks** [comment]
@amerkoleci  not had a chance in the last week to do any more progress, work in RL is basically the major item.  Im going to look at some changes to the code gen and structure. I still do not understand how to get the async working and I havent had a chance to test fully.

**Comment by Gawson** [comment]
@h1cks I'll check your repo. In the meantime I'm working on C++ code to make IAsynCallback implementation work, next I'll try to port it to Vortice. If I have any working code I'll share it. The tricky part is our project is in .net5 but with C++/WinRT runtime component connected with CsWinRT for some native helpers not present in any C# lib (Win2D <> Direct2D1 interop) so making custom com object is sometimes problematic.

**Comment by Gawson** [comment]
@h1cks I have working callback example in C++/WinRT. It appears that WinRT greatly simplifies implementing COM interfaces.
Most of the COM heavy lifting is done by `implements<>` template. Rest of unimplemented pure virtuals may be generated from IMFAsyncCallback - only crucial methods will be generated without QueryInterface, AddRef and rest of the COM specific ones.

`CAsyncCallback.h`
```cpp
#pragma once
#include <mfapi.h>

namespace winrt::NativeInteropRT::implementation
{
	struct CAsyncCallback : implements<CAsyncCallback, IMFAsyncCallback> {

		CAsyncCallback() {
			
		}

		~CAsyncCallback() {

			OutputDebugString(L"Destroying callback");
		}

		// Inherited via implements
		virtual HRESULT __stdcall GetParameters(DWORD* pdwFlags, DWORD* pdwQueue) override;

		virtual HRESULT __stdcall Invoke(IMFAsyncResult* pAsyncResult) override;

		virtual void use_make_function_to_create_this_object() override;

	};

}
```
`CAsyncCallback.cpp`
```cpp
#include "pch.h"
#include "CAsyncCallback.h"

HRESULT __stdcall winrt::NativeInteropRT::implementation::CAsyncCallback::GetParameters(DWORD* pdwFlags, DWORD* pdwQueue)
{

	return 0;
}

HRESULT __stdcall winrt::NativeInteropRT::implementation::CAsyncCallback::Invoke(IMFAsyncResult* pAsyncResult)
{
	OutputDebugString(L"INVOKED!!!!!");


	return 0;
}

void winrt::NativeInteropRT::implementation::CAsyncCallback::use_make_function_to_create_this_object()
{
}

```

and simple usage example based on PresentationClock:

```cpp
{
            com_ptr<IMFPresentationClock> clock;
            HRESULT success = MFCreatePresentationClock(clock.put());

            com_ptr<IMFPresentationTimeSource> timeSource;
            success = MFCreateSystemTimeSource(timeSource.put());

            clock->SetTimeSource(timeSource.get());

            com_ptr<CAsyncCallback> callback;
            CAsyncCallback call = CAsyncCallback();
            callback.attach(&call);

            com_ptr<IMFTimer> timer = clock.as<IMFTimer>();

            com_ptr<IUnknown> cancelationToken;
            success = timer->SetTimer(0, 60000000, callback.as<IMFAsyncCallback>().get(), NULL, cancelationToken.put());

            clock->Start(0);


            auto lambda = [=](int a) { 
                int counter = a;
                while (counter--) {
                    Sleep(100);
                    MFTIME time;
                    clock->GetTime(&time);
                    OutputDebugString(to_hstring(time).c_str());
                    OutputDebugString(L"\n");
                }
            };

            std::thread thread(lambda, 10);
            thread.join();            
            
            clock->Stop();
        }
```
In the sample when timer reaches 60000000 callback is invoked.
As this is in the runtime component I'll check if CsWinRT wrapper can be used with MF wrappers on the C# side. Generated classes has internal native pointers and implements necessary COM interfaces so it should work.

@amerkoleci @h1cks MediaFoundation code lacks of some crucial GUIDS like MFVideoFormat_H264. There is MFVideoFormat_H264_ES but there is no proper GUID for plain H264. I've checked in lib and in the source code. Maybe I'll be able to list all differences between SharpDX and Vortice this week.

**Comment by h1cks** [comment]
I sadly will be a while before I have time to continue to work on the code.  Feel free to do a code submission for the GUIDs.  My work, RL and other pieces have taken me away from working on this.  I literally have no spare time atm.

**Comment by amerkoleci** [comment]
Please write me down an example your would like to port from C++ and guids that are missing and I will add them starting from tomorrow.

**Comment by h1cks** [comment]
> Please write me down an example your would like to port from C++ and guids that are missing and I will add them starting from tomorrow.

@amerkoleci  apologies I have not been on much, but my actual real life job has taken most of my time at the moment.

**Comment by amerkoleci** [comment]
> > Please write me down an example your would like to port from C++ and guids that are missing and I will add them starting from tomorrow.
> 
> @amerkoleci apologies I have not been on much, but my actual real life job has taken most of my time at the moment.

Don't worry, we're doing our best :)

**Comment by amerkoleci** [comment]
I will close this issues as it's getting to much big, let's open issues with given sub topic/request
